src/EditorFeatures/CSharpTest/CodeActions/ConvertLinq/ConvertForEachToLinqQueryTests.cs

```diff
@@ -0,0 +1,2842 @@
+        public async Task CountInMultipleDeclarationNonZeroToReturnNotLast()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt = 5, c = 0;
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                cnt++;
+            }
+        }|]
+
+        return cnt;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt = 5, c = 0;
+        cnt += (from int n1 in nums
+                from int n2 in nums
+                select n1).Count();
+        return cnt;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task CountInAssignmentToZero()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt;
+        cnt = 0;
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                cnt++;
+            }
+        }|]
+
+        return cnt;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt;
+        return (from int n1 in nums
+                from int n2 in nums
+                select n1).Count();
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task CountInAssignmentToNonZero()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt;
+        cnt = 5;
+        [|foreach (int n1 in nums)
+        {
+            foreach (int n2 in nums)
+            {
+                cnt++;
+            }
+        }|]
+
+        return cnt;
+    }
+}
+";
+            string output = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums)
+    {
+        int cnt;
+        cnt = 5;
+        cnt += (from int n1 in nums
+                from int n2 in nums
+                select n1).Count();
+        return cnt;
+    }
+}
+";
+
+            await TestInRegularAndScriptAsync(source, output);
+        }
+
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsConvertForEachToQuery)]
+        public async Task CountInParameterAssignedToZeroAndReturned()
+        {
+            string source = @"
+using System.Collections.Generic;
+using System.Linq;
+class C
+{
+    int M(IEnumerable<int> nums, int c)
+    {
+        c = 0;
```

Comment:
these one. I guess unlike yield return one, doesn't care whether there is more statements before foreach in the method?

Reply:
yes, we check that the assignment is performed just before the foreach. We do not care about previous statements.

---
In reply to: [195859157](https://github.com/dotnet/roslyn/pull/27130#discussion_r195859157) [](ancestors = 195859157)